### PR TITLE
trace: Do not produce dead code calling BIO_printf if disabled

### LINKED ISSUE #15880: Windows fips debug build fails with "unresolved external symbol _BIO_printf"
Windows build with `enable-fips` and `--debug` options fails with `error LNK2001: unresolved external symbol _BIO_printf`.  The build succeeds if either `enable-fips` or `--debug` is not included.

This is new with the 3.0 beta1 release.  It was not an issue in the alpha17 release.

```
perl Configure VC-WIN32 enable-fips --debug
nmake
...
providers\fips.def(5) : warning LNK4093: Drive/Directory component ignored in 'LIBRARY' statement
   Creating library providers\fips.lib and object providers\fips.exp
fips.exp : warning LNK4070: /OUT:fips-3.dll directive in .EXP differs from output filename 'providers\fips.dll'; ignoring directive
libfips.lib(libfips-lib-ec_key.obj) : error LNK2001: unresolved external symbol _BIO_printf
libfips.lib(libfips-lib-ecp_nistz256.obj) : error LNK2001: unresolved external symbol _BIO_printf
libfips.lib(libfips-lib-ec_mult.obj) : error LNK2001: unresolved external symbol _BIO_printf
libfips.lib(libfips-lib-p_lib.obj) : error LNK2001: unresolved external symbol _BIO_printf
libfips.lib(libfips-lib-dh_lib.obj) : error LNK2001: unresolved external symbol _BIO_printf
libfips.lib(libfips-lib-dsa_lib.obj) : error LNK2001: unresolved external symbol _BIO_printf
libfips.lib(libfips-lib-rsa_lib.obj) : error LNK2001: unresolved external symbol _BIO_printf
libfips.lib(libfips-lib-ecx_key.obj) : error LNK2001: unresolved external symbol _BIO_printf
providers\fips.dll : fatal error LNK1120: 1 unresolved externals
Could Not Find D:\work\crypto\openssl-3.0.0-beta1\fips.*
NMAKE : fatal error U1077: 'cmd' : return code '0x1'
Stop.
NMAKE : fatal error U1077: '"C:\Program Files (x86)\Microsoft Visual Studio\2017\Professional\VC\Tools\MSVC\14.12.25827\bin\HostX86\x86\nmake.exe"' : return code '0x2'
Stop.
```
```
perl configdata.pm --dump

Command line (with current working directory = .):

    C:\Strawberry\perl\bin\perl.exe Configure VC-WIN32 enable-fips --debug 

Perl information:

    C:\Strawberry\perl\bin\perl.exe
    5.30.2 for MSWin32-x64-multi-thread

Enabled features:

    aria
    asm
    async
    autoalginit
    autoerrinit
    autoload-config
    bf
    blake2
    bulk
    cached-fetch
    camellia
    capieng
    cast
    chacha
    cmac
    cmp
    cms
    comp
    ct
    deprecated
    des
    dgram
    dh
    dsa
    dso
    dtls
    dynamic-engine
    ec
    ec2m
    ecdh
    ecdsa
    engine
    err
    filenames
    fips
    fips-securitychecks
    gost
    idea
    legacy
    loadereng
    makedepend
    md4
    mdc2
    module
    multiblock
    nextprotoneg
    ocb
    ocsp
    padlockeng
    pic
    pinshared
    poly1305
    posix-io
    psk
    rc2
    rc4
    rdrand
    rfc3779
    rmd160
    scrypt
    secure-memory
    seed
    shared
    siphash
    siv
    sm2
    sm3
    sm4
    sock
    srp
    srtp
    sse2
    ssl
    ssl-trace
    static-engine
    stdio
    tests
    threads
    tls
    ts
    ui-console
    uplink
    whirlpool
    tls1
    tls1-method
    tls1_1
    tls1_1-method
    tls1_2
    tls1_2-method
    tls1_3
    dtls1
    dtls1-method
    dtls1_2
    dtls1_2-method

Disabled features:

    acvp-tests          [default]   OPENSSL_NO_ACVP_TESTS
    afalgeng            [not-linux] OPENSSL_NO_AFALGENG
    asan                [default]   OPENSSL_NO_ASAN
    buildtest-c++       [default]
    crypto-mdebug       [default]   OPENSSL_NO_CRYPTO_MDEBUG
    devcryptoeng        [default]   OPENSSL_NO_DEVCRYPTOENG
    ec_nistp_64_gcc_128 [default]   OPENSSL_NO_EC_NISTP_64_GCC_128
    egd                 [default]   OPENSSL_NO_EGD
    external-tests      [default]   OPENSSL_NO_EXTERNAL_TESTS
    fuzz-afl            [default]   OPENSSL_NO_FUZZ_AFL
    fuzz-libfuzzer      [default]   OPENSSL_NO_FUZZ_LIBFUZZER
    ktls                [default]   OPENSSL_NO_KTLS
    md2                 [default]   OPENSSL_NO_MD2 (skip crypto\md2)
    msan                [default]   OPENSSL_NO_MSAN
    rc5                 [default]   OPENSSL_NO_RC5 (skip crypto\rc5)
    sctp                [default]   OPENSSL_NO_SCTP
    trace               [default]   OPENSSL_NO_TRACE
    ubsan               [default]   OPENSSL_NO_UBSAN
    unit-test           [default]   OPENSSL_NO_UNIT_TEST
    weak-ssl-ciphers    [default]   OPENSSL_NO_WEAK_SSL_CIPHERS
    zlib                [default]
    zlib-dynamic        [default]
    ssl3                [default]   OPENSSL_NO_SSL3
    ssl3-method         [default]   OPENSSL_NO_SSL3_METHOD

Config target attributes:

    AR => "lib",
    ARFLAGS => "/nologo",
    AS => "nasm",
    ASFLAGS => "",
    CC => "cl",
    CFLAGS => "/W3 /wd4090 /nologo /Od",
    CPP => "\"\$(CC)\" /EP /C",
    HASHBANGPERL => "/usr/bin/env perl",
    LD => "link",
    LDFLAGS => "/nologo /debug",
    MT => "mt",
    MTFLAGS => "-nologo",
    RANLIB => "CODE(0x2795c88)",
    RC => "rc",
    ar_resp_delim => "
",
    aroutflag => "/out:",
    asflags => "-f win32",
    asm_arch => "x86",
    asoutflag => "-o ",
    bin_cflags => "/Zi /Fdapp.pdb /MDd",
    bin_lflags => "setargv.obj /subsystem:console /opt:ref",
    bn_ops => "BN_LLONG",
    build_file => "makefile",
    build_scheme => [ "unified", "windows", "VC-common", "VC-WOW" ],
    cflags => "/Gs0 /GF /Gy /MDd",
    coutflag => "/Fo",
    cppflags => "",
    defines => [ "OPENSSL_BUILDING_OPENSSL", "OPENSSL_SYS_WIN32", "WIN32_LEAN_AND_MEAN", "UNICODE", "_UNICODE", "_CRT_SECURE_NO_DEPRECATE", "_WINSOCK_DEPRECATED_NO_WARNINGS", "DEBUG", "_DEBUG" ],
    disable => [  ],
    dso_cflags => "/Zi /Fddso.pdb",
    dso_scheme => "win32",
    enable => [  ],
    ex_libs => "ws2_32.lib gdi32.lib advapi32.lib crypt32.lib user32.lib",
    includes => [  ],
    ld_implib_flag => "/implib:",
    ld_resp_delim => "
",
    lddefflag => "/def:",
    ldoutflag => "/out:",
    ldpostoutflag => "",
    ldresflag => " ",
    lflags => "",
    lib_cflags => "/Zi /Fdossl_static.pdb",
    lib_cppflags => "",
    lib_defines => [ "L_ENDIAN" ],
    makedep_scheme => "VC",
    makedepcmd => "\"\$(CC)\" /Zs /showIncludes",
    module_cflags => "",
    module_cxxflags => undef,
    module_ldflags => "/dll",
    mtinflag => "-manifest ",
    mtoutflag => "-outputresource:",
    perl_platform => "Windows::MSVC",
    perlasm_scheme => "win32n",
    rcoutflag => "/fo",
    shared_cflag => "",
    shared_defflag => "",
    shared_defines => [  ],
    shared_ldflag => "/dll",
    shared_rcflag => "",
    shared_target => "win-shared",
    sys_id => "WIN32",
    thread_defines => [  ],
    thread_scheme => "winthreads",
    unistd => "<unistd.h>",
    uplink_arch => "common",

Recorded environment:

    AR =
    ARFLAGS =
    AS =
    ASFLAGS =
    BUILDFILE =
    CC =
    CFLAGS =
    CPP =
    CPPDEFINES =
    CPPFLAGS =
    CPPINCLUDES =
    CROSS_COMPILE =
    CXX =
    CXXFLAGS =
    HASHBANGPERL =
    LD =
    LDFLAGS =
    LDLIBS =
    MT =
    MTFLAGS =
    OPENSSL_LOCAL_CONFIG_DIR =
    PERL =
    RANLIB =
    RC =
    RCFLAGS =
    RM =
    WINDRES =
    __CNF_CFLAGS =
    __CNF_CPPDEFINES =
    __CNF_CPPFLAGS =
    __CNF_CPPINCLUDES =
    __CNF_CXXFLAGS =
    __CNF_LDFLAGS =
    __CNF_LDLIBS =

Makevars:

    AR              = lib
    ARFLAGS         = /nologo
    AS              = nasm
    CC              = cl
    CFLAGS          = /W3 /wd4090 /nologo /Od
    CPP             = "$(CC)" /EP /C
    CPPDEFINES      =
    CPPFLAGS        =
    CPPINCLUDES     =
    CXXFLAGS        =
    HASHBANGPERL    = /usr/bin/env perl
    LD              = link
    LDFLAGS         = /nologo /debug
    LDLIBS          =
    MT              = mt
    MTFLAGS         = -nologo
    PERL            = C:\Strawberry\perl\bin\perl.exe
    RANLIB          = ranlib
    RC              = rc
    RCFLAGS         =

NOTE: These variables only represent the configuration view.  The build file
template may have processed these variables further, please have a look at the
build file for more exact data:
    makefile

build file:

    makefile

build file templates:

    Configurations\common0.tmpl
    Configurations\windows-makefile.tmpl
```

### PR DESCRIPTION
**Pull request by t8m**
Fixes #15880

### CODE REVIEW

**Review by levitte — APPROVED** [review]

**Comment by levitte** [comment]
The approval is, of course, subject to CI success

File: include/openssl/trace.h [hunk 1]
```diff
@@ -270,11 +270,19 @@ void OSSL_trace_end(int category, BIO *channel);
...
+# if !defined OPENSSL_NO_TRACE && !defined FIPS_MODULE
+
+#  define OSSL_TRACEV(category, args) \
     OSSL_TRACE_BEGIN(category) \
         BIO_printf args; \
     OSSL_TRACE_END(category)
 
+# else
+
+#  define OSSL_TRACEV(category, args) ((void)0)
```
Comment (paulidale) [inline include/openssl/trace.h:282]: Why the `((void)0)`?  Leaving the body empty should work.

**Review by paulidale — APPROVED** [review]

File: include/openssl/trace.h (hunk 1 above)
Comment (levitte) [reply include/openssl/trace.h:282]: Not necessarily...  some compilers warn about extra `;`, which would be the consequence of an empty body

Comment (paulidale) [reply include/openssl/trace.h:282]: Silly compilers ...

**Comment by t8m** [comment]
Merged to master. Thank you for the reviews.

**Pull request closed without merging**